### PR TITLE
Add cloud function for generating patient hash

### PR DESCRIPTION
**Pull request by nickclyde**
Resolves https://app.clickup.com/t/2q4v004

Adds the python code for the cloud function as well as the necessary terraform config. Also added a secret to be stored in Secret Manager for `PATIENT_HASH_SALT` which is then loaded into the cloud function as an environment variable.

### CODE REVIEW

File: cloud-functions/add_patient_hash/main.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            # Perform the phone standardization
+            request_json = request.get_json(silent=False)
+            salt_str = os.environ.get("PATIENT_HASH_SALT")
+            final_response = add_patient_identifier(request_json, salt_str)
+        else:
+            return body_response
+    else:
+        return header_response
+
+    return final_response
```
Comment (DanPaseltiner) [inline cloud-functions/add_patient_hash/main.py:43]: What do you think about the following changes? From my perspective they accomplish:

1.  Eliminates the need for the `else` statements.
2. Handles the case when there is no `PATIENT_HASH_SALT` env variable. 

To do this we would need to add a `check_for_environment_variable` function to our utils module. I'm currently using some environment variables in the PR for the `read_source_data` function and could use the function there as well.

```suggestion
    if header_response.status_code == 400:
        return header_response

    body_response = validate_fhir_bundle_or_resource(request)
    if body_response.status_code == 400:
        return body_response

    environement_check_response = check_for_environment_variable("PATIENT_HASH")
    if environement_check_response.status_code == 400:
        return environement_check_response

     # Perform the phone standardization
    request_json = request.get_json(silent=False)
    salt_str = os.environ.get("PATIENT_HASH_SALT")
    final_response = add_patient_identifier(request_json, salt_str)
    return final_response
```

**Review by DanPaseltiner — CHANGES_REQUESTED** [review]
Nice work! One minor comment on documentation and one other that's a little more substantive. I might be missing something and we certainly don't have to do things the way I suggested. Happy to chat further if you like.

File: cloud-functions/add_patient_hash/main.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    content_type = "application/json"
+    # Validate request header.
+    header_response = validate_request_header(request, content_type)
+
+    # Check that the request body contains a FHIR bundle or resource.
+    if header_response.status_code != 400:
+        body_response = validate_fhir_bundle_or_resource(request)
+
+        if body_response.status_code != 400:
+            # Perform the phone standardization
```
Comment (BradySkylight) [inline cloud-functions/add_patient_hash/main.py:34]: This looks like a left over comment from copy and paste.  I also agree with Dan's suggestion about the IF ELSE changes.  If you look at my latest PR for the Phone Standardization I think you will see an example of what I think he is referring to.

File: cloud-functions/add_patient_hash/main.py (hunk 1 above)
Comment (nickclyde) [reply cloud-functions/add_patient_hash/main.py:43]: This looks good! I'll go ahead and add that function and incorporate these changes 👍

File: cloud-functions/add_patient_hash/main.py (hunk 2 above)
Comment (nickclyde) [reply cloud-functions/add_patient_hash/main.py:34]: Good catch, I'll remove this!

**Comment by celestemespinoza** [comment]
Task linked: [CU-2q4v004 Write GCP patient hash function](https://app.clickup.com/t/2q4v004)

File: cloud-functions/phdi_cloud_function_utils/phdi_cloud_function_utils/__init__.py [hunk 3]
```diff
@@ -84,3 +85,22 @@ def validate_fhir_bundle_or_resource(request: Request) -> Response:
...
+    for environment_variable in environment_variables:
+        if os.environ.get(environment_variable) is None:
+            error_message = (
+                "Environment variable '{}' not set. "
+                + "The environment variable must be set."
+            ).format(environment_variable)
+            logging.error(error_message)
+            return fail(message=error_message, status="Bad Request")
+
+    return success()
```
[the remark this answers is not in the document]
Comment (DanPaseltiner) [reply cloud-functions/phdi_cloud_function_utils/phdi_cloud_function_utils/__init__.py:106]: PR 37 now includes these changes.

**Review by DanPaseltiner — APPROVED** [review]
Thanks for putting this together Nick! I think there are a couple places to realign with `main` after PR 37 lands, but I don't want a review from to be blocking.

**Pull request merged**